### PR TITLE
[WORKFLOWS-181] Create and add teams to Tower workspaces

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
We found that Tower users weren't getting access to the right workspaces. As a first-pass attempt to resolving this issue, I'm updating the Tower configuration script to create teams per project per role and adding these teams to the relevant workspaces. I'm also performing the additional step of removing unexpected users from each team to make sure they remain in sync with the Sceptre config file. 

I was able to test the updated script in Tower-dev. Once it's deployed to Tower-prod, I'm going to remove all of the users that are listed individually on workspaces. Because these users will have been added via a team, no one should suffer a lapse of access (unless they weren't supposed to have access to the workspace in the first place). 

**Edit:** As I show in [this comment](https://github.com/Sage-Bionetworks-Workflows/nextflow-infra/pull/96#issuecomment-1064220965), workspace participants with the `maintainer` role can add people to the workspace. I don't know if we can revoke this permission without affecting their ability to launch workflows. At the moment, the script ensures that the team membership is in sync with the Sceptre config file. Should we also make sure that no one else is added to the workspace as individual users (_i.e._ outside of a team)? 

## Organization Teams 

image

## Example Team

image

## Teams Added to Workspace

image

### CODE REVIEW

File: bin/configure-tower-projects.py
Comment (thomasyu888) [inline bin/configure-tower-projects.py:386]: Nit: This function now adds both participant and team? Ok with function name but improve docs?

File: bin/configure-tower-projects.py
Comment (thomasyu888) [inline bin/configure-tower-projects.py:415]: nit: Why not user_id when its team_id?

File: bin/configure-tower-projects.py
Comment (thomasyu888) [inline bin/configure-tower-projects.py:727]: Nit: I saw this `/orgs/{orgId}/teams/validate`: Check that is a valid team name. 

Does this mean check if team exists? Or if the actual team name is compliant with nextflow tower.

File: bin/configure-tower-projects.py
Comment (thomasyu888) [inline bin/configure-tower-projects.py:750]: Nit: use list comprehension
```suggestion
        team_member_ids = [
            team_member["memberId"]
            for team_member in response["members"]
        ]
```

File: bin/configure-tower-projects.py
Comment (thomasyu888) [inline bin/configure-tower-projects.py:796]: Nit: It is more apparent after reading `__init__` that on the instantiation of the TowerOrganization class that both `self.create()` and `self.populate()` is called?  I can't really think of a scenario when `create_workspaces` would be called without the instantiation of the class.

File: bin/configure-tower-projects.py
Comment (thomasyu888) [inline bin/configure-tower-projects.py:759]: Apologies, as Im not entirely familiar, but where are these users for projects specified?

File: README.md
Comment (thomasyu888) [inline README.md:58]: Are we the only admins?

**Review by thomasyu888 — COMMENTED** [review]
Clever approach - I like how you make sure that the config file is the source of truth.  That said - are members of teams allowed to add users to teams in the UI?  Does it become an issue if people are adding members to teams and this stack is re-run.

**Comment by BrunoGrandePhD** [comment]
@thomasyu888: The `nextflow-tower-admin` service account has the `owner` role on the Sage organization and can modify team membership. Everyone else just has the `member` role, which has limited read-only access to the organization. Here's what I see on the organization page with my own account:

image

Contrast that with what the owner sees:

image

That said, workspace maintainers (_i.e._ the users listed under `S3ReadWriteAccessArns`) can add users to a workspace. When this is done, the user is added with the `launch` role, which is only one step above the `view` role. Also, being added to the workspace doesn't grant you access to the bucket, so it's of limited usefulness and risk. 

image

A legitimate question is whether we also want to clean up the workspace participants to make sure everything stay in sync with the Sceptre config file, which would be the ground truth. I'm leaning towards yes, but I'm open to thoughts.

File: README.md
Comment (BrunoGrandePhD) [reply README.md:58]: The `nextflow-tower-admin` service account is the only owner on the organization, teams, and workspaces. This way, it's harder for the Tower setup to drift from what this script deploys.

File: bin/configure-tower-projects.py
Comment (BrunoGrandePhD) [reply bin/configure-tower-projects.py:727]: It does alert you when the team name already exists, and it also checks if the team name contains valid characters. The former case is handled with the existing code, and the latter case shouldn't happen because we're appending `-<role>` to the stack name, which already only consists of valid characters. 

I could use the dedicated `/validate` endpoint to check if a team name already exists. Can you think of any benefits? 

```
>>> tower.request("GET", "/orgs/193738885884191/teams/validate?name=example-dev-project-maintainers")
{'message': "A team with name 'example-dev-project-maintainers' already exists"}

>>> tower.request("GET", "/orgs/193738885884191/teams/validate?name=#winning")
{'message': 'Team name must contain a minimum of 1 and a maximum of 38 alphanumeric characters separated by dashes or underscores'}
```

File: bin/configure-tower-projects.py
Comment (BrunoGrandePhD) [reply bin/configure-tower-projects.py:415]: If something were to be `user_id`, it would be the `member_id`. Currently, `user` contains emails because that's what's stored in the Sceptre config files. I'm not using `email` as the variable name because it can technically contain a username (as indicated by the Tower field, `userNameOrEmail`). I'm not particularly eager to use `user_name_or_email` either. 

Can you think of a better variable name?

Comment (BrunoGrandePhD) [reply bin/configure-tower-projects.py:415]: I could refactor the code so that member IDs are tracked as soon as they're obtained instead of relying on `self.org.members` to retrieve the member ID in this function.

File: bin/configure-tower-projects.py
Comment (BrunoGrandePhD) [reply bin/configure-tower-projects.py:386]: This isn't a nit. The docs were inaccurate, so I've updated them. "Participant" is the term that Tower uses to refer to users or teams added to a workspace.

File: bin/configure-tower-projects.py
Comment (thomasyu888) [reply bin/configure-tower-projects.py:415]: I see. This makes sense. No need to refactor, user makes sense here.

File: bin/configure-tower-projects.py
Comment (thomasyu888) [reply bin/configure-tower-projects.py:727]: Are the teams scoped to the tower project? The efficiency would be not having to loop through a list of teams as the team list grows but it's really not too big of a burden

File: bin/configure-tower-projects.py
Comment (thomasyu888) [reply bin/configure-tower-projects.py:386]: Agreed, the docstring for this function still seems unchanged.

Comment (BrunoGrandePhD) [reply bin/configure-tower-projects.py]: Sorry, I forgot to upload my commit.

**Review by thomasyu888 — APPROVED** [review]
LGTM : up to you if you want to make the change with the validation of the team name.  Even with hundreds of teams, it won't be computationally expensive.

File: bin/configure-tower-projects.py
Comment (BrunoGrandePhD) [reply bin/configure-tower-projects.py:727]: > Are the teams scoped to the tower project?

There is one team per project per role (_i.e._ read/write aka maintainer and read-only aka viewer). These teams exist under the organization and aren't tied to the project workspace until they're added to the workspace as participants with the relevant role. I'm not sure if this answers your question. 

We'll need to iterate through the teams anyways to make sure their membership is up-to-date. We can see what the runtime is once we deploy to production. The dev environment only has two teams, so it's hard to tell.

**Pull request merged**
